flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/DefaultSlotAssigner.java

```diff
@@ -22,56 +22,119 @@
-                        });
-        return sharedSlotToVertexAssignment.values().stream()
-                .map(ExecutionSlotSharingGroup::new)
-                .collect(Collectors.toList());
+    private Collection<? extends SlotInfo> pickSlotsIfNeeded(
+            int requestExecutionSlotSharingGroups, Collection<? extends SlotInfo> freeSlots) {
+        Collection<? extends SlotInfo> pickedSlots = freeSlots;
+        // To avoid the sort-work loading.
+        if (APPLICATION_MODE_EXECUTION_TARGET.equalsIgnoreCase(executionTarget)
+                && minimalTaskManagerPreferred
+                && freeSlots.size() > requestExecutionSlotSharingGroups) {
+            final Map<TaskManagerLocation, ? extends Set<? extends SlotInfo>> slotsPerTaskExecutor =
+                    getSlotsPerTaskExecutor(freeSlots);
+            pickedSlots =
+                    pickSlotsInMinimalTaskExecutors(
+                            slotsPerTaskExecutor, requestExecutionSlotSharingGroups);
+        }
+        return pickedSlots;
+    }
+
+    /**
+     * In order to minimize the using of task executors at the resource manager side in the
+     * application-mode and release more task executors in a timely manner, it is a good choice to
+     * prioritize selecting slots on task executors with the most available slots.
+     */
+    private Iterator<TaskManagerLocation> getSortedTaskExecutors(
+            Map<TaskManagerLocation, ? extends Set<? extends SlotInfo>> slotsPerTaskExecutor) {
+        final Comparator<TaskManagerLocation> taskExecutorComparator =
+                (leftTml, rightTml) ->
+                        Integer.compare(
+                                slotsPerTaskExecutor.get(rightTml).size(),
+                                slotsPerTaskExecutor.get(leftTml).size());
+        return slotsPerTaskExecutor.keySet().stream().sorted(taskExecutorComparator).iterator();
+    }
+
+    /**
+     * Pick the target slots to assign with the requested groups.
+     *
+     * @param slotsByTaskExecutor slots per task executor.
+     * @param requestedGroups the number of the request execution slot sharing groups.
+     * @return the target slots that are distributed on the minimal task executors.
+     */
+    private Collection<? extends SlotInfo> pickSlotsInMinimalTaskExecutors(
+            Map<TaskManagerLocation, ? extends Set<? extends SlotInfo>> slotsByTaskExecutor,
+            int requestedGroups) {
+        final List<SlotInfo> pickedSlots = new ArrayList<>();
+        final Iterator<TaskManagerLocation> sortedTaskExecutors =
+                getSortedTaskExecutors(slotsByTaskExecutor);
+        while (pickedSlots.size() < requestedGroups) {
+            Set<? extends SlotInfo> slotInfos = slotsByTaskExecutor.get(sortedTaskExecutors.next());
+            pickedSlots.addAll(slotInfos);
+        }
+        return pickedSlots;
+    }
+
+    static Map<TaskManagerLocation, ? extends Set<? extends SlotInfo>> getSlotsPerTaskExecutor(
```

Comment:
Can be private. Why this one is `static` and `getSortedTaskExecutors`, `pickSlotsInMinimalTaskExecutors` are not? All 3 can be `static`, either make all of them `static` or none.

Reply:
Changed its as non-static-private.